### PR TITLE
Ensure use of generateIdIfAbsentFromDocument assigns the result

### PR DESCRIPTION
**Pull request by rozza**
There is no guarantee that CollectibleCodec<TDocument> will be able to
mutate the TDocument instance.

JAVA-2272

### CODE REVIEW

**Comment by rozza** [comment]
This had already been done in the sync version, so just the async usages needed updating.

**Comment by rozza** [comment]
Added regression test cases

File: driver-core/src/test/unit/com/mongodb/client/ImmutableDocument.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @Override
+    public void clear() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return immutableDocument.keySet();
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/client/ImmutableDocument.java]: True, but it's test code so doesn't really matter if it's fully immutable.

**Comment by jyemin** [comment]
LGTM

File: driver-core/src/test/unit/com/mongodb/client/ImmutableDocument.java (hunk 1 above)
Comment (rozza) [inline driver-core/src/test/unit/com/mongodb/client/ImmutableDocument.java]: Don't think you can - I just wrapping the `Collections.unmodifiableMap` which returns unmodifiable sets and unmodifiable entries.

**Pull request merged**